packages/react/src/components/form/authenticating/totp.tsx

```diff
@@ -0,0 +1,283 @@
+  const { values, registerField, registerSubmit, setError, clearError } =
+    useForm();
+  const { title, message, submit } = getTextKeys(flowState);
+  const uiState = flowState.getChildState();
+  const [showUri, setShowUri] = useState(false);
+  const [useRecoveryCode, setUseRecoveryCode] = useState(false);
+  const submitInputRef = useRef<HTMLInputElement>(null);
+
+  useEffect(() => {
+    const handler = () => {
+      setError("totp", {
+        message: text["authenticating.otpInput.submit.error"],
+      });
+      values["totp"] = "";
+    };
+    sid?.subscribe("otpIncorrectCodeSubmitted", handler);
+
+    return () => sid?.unsubscribe("otpIncorrectCodeSubmitted", handler);
+  }, [setError, sid, text, values]);
+
+  const handleChange = useCallback(
+    (otp: string) => {
+      const onChange = registerField("totp", {
+        validator: (value) => {
+          if (!useRecoveryCode && !isValidOTPCode(value)) {
+            return { message: text["validationError.otp"] };
+          }
+        },
+      });
+      const event = {
+        target: {
+          value: otp,
+        },
+      };
+
+      clearError("totp");
+      onChange(event as never);
+    },
+    [clearError, registerField, text, useRecoveryCode]
+  );
+
+  useEffect(() => {
+    if (isValidOTPCode(values["totp"])) {
+      // Automatically submit the form when the TOTP code is valid
+      submitInputRef.current?.click();
+    }
+  }, [values]);
+
+  return (
+    <>
+      <BackButton onCancel={() => flowState.cancel()} />
+      <Text as="h1" t={title} variant={{ size: "2xl-title", weight: "bold" }} />
+      <Text t={message} variant={{ color: "contrast", weight: "semibold" }} />
+      {flowState.matches("initial") && <Loader />}
+      {isRegisterTotpAuthenticatorState(uiState) && (
+        <form
+          onSubmit={registerSubmit((e) => {
+            e.preventDefault();
+            uiState.confirm();
+          })}
+          className={styles.registerAuthenticatorForm}
+        >
+          <img
+            src={`data:image/png;base64, ${uiState.qrCode}`}
+            alt="QR Code"
+            className={styles.qrCode}
+          />
+          <Prompt
+            prompt="authenticating.registerAuthenticator.totp.prompt"
+            cta="authenticating.registerAuthenticator.totp.cta"
+            onClick={() => setShowUri(true)}
+          />
+          {showUri && (
+            <ReadOnly
+              id="uri"
+              value={uiState.uri}
+              label="Set up URI"
+              copy
+              className={styles.readOnly}
+            />
+          )}
+          <Submit textKey={submit} />
+        </form>
+      )}
+      {isInputState(uiState) && (
+        <form
+          onSubmit={registerSubmit((e) => {
+            e.preventDefault();
+            uiState.submit(values["totp"]);
+          })}
+          className={styles.otpForm}
+        >
+          <div className={styles.formInner}>
+            {useRecoveryCode ? (
```

Comment:
Note: The design has a gap, there is no screen for logging in with a recovery code, something I overlooked. I figured that it's not super hard to do this now, without reiterating on the design. Let me know WDYT, I think it's sufficient and most importantly - it works. Maybe we can polish it a bit more if needed ✨